### PR TITLE
[#1367] First-class warranty tracking on commodities

### PR DESCRIPTION
**Pull request by denisvmedia**
## Summary

Implements first-class warranty tracking on commodities (#1367) end-to-end — backend data model + worker + email + filters, FE form/detail/list view, and a chromium-verified e2e walking the acceptance criterion.

### Backend

- New `warranty_expires_at` + `warranty_notes` columns on `commodities`. Status (`active`/`expiring`/`expired`/`none`) is computed via `models.ComputeWarrantyStatus` with a 60-day "expiring" window (`models.WarrantyExpiringWindowDays`).
- New `warranty_reminders` table keyed `(commodity_id, threshold_days)` — the worker's idempotency store. Wired into `GroupPurger` so a hard-deleted group leaves no orphan rows.
- `WarrantyReminderRegistry` (memory + postgres). Postgres impl uses `INSERT … ON CONFLICT DO NOTHING` so the worker's race-loser path is the same as the already-sent path.
- `WarrantyReminderService` + `WarrantyReminderWorker` — single-goroutine ticker (default 1h, configurable via `--warranty-reminder-interval`). Fires reminders at 60/30/7 days. Email recipients are group admins; falls back to `created_by_user_id` when memberships are empty (single-user installs).
- New email template (`warranty_reminder.{html,txt}.tmpl`) wired through `AsyncEmailService.SendWarrantyReminderEmail` and the `EmailService` interface.
- `GET /api/v1/g/{slug}/commodities` accepts `warranty_status=active|expiring|expired|none` (multi) and `warranty_expires_before=YYYY-MM-DD`.

### Frontend

- Commodity form's Warranty step now has expiry-date + notes inputs with a live status pill preview (replaces the previous "Coming soon" banner).
- Commodity detail's Warranty tab renders status pill, days-remaining counter, and notes. Empty state nudges toward the edit dialog.
- Dedicated `/g/:slug/warranties` page — group-wide list with **All / Active / Expiring / Expired / None** tabs. Each non-`all` tab maps directly to the BE `warranty_status=` filter; `all` filters client-side to "rows with any tracked warranty". Sidebar entry was already wired.
- New `warranties` i18n namespace (en) + cs/ru stubs; preserve patterns added for the dynamic tab/empty/status keys so `npm run i18n:check` is clean.
- Regenerated swagger + `api.d.ts` so the FE picks up the new fields.

### Tests

- `services.TestWarrantyReminderService_RemindOnce_TickClock` pins the time-based acceptance criterion: 65-day expiry → status active; tick clock 35 days forward → status expiring + reminder fires; re-run at the same clock → idempotent (no second email).
- `models.TestComputeWarrantyStatus` covers boundary semantics for the computed status.
- `registry/memory.TestCommodityRegistry_ListPaginated_WarrantyFilter` pins the new `warranty_status` + `warranty_expires_before` query parameters.
- New `frontend WarrantiesListPage.test.tsx` (4 cases: empty state, row rendering, tab toggle, deep-link preselection).
- New `e2e/tests/warranties.spec.ts` walks the FE flow on chromium: seed three commodities (65d/30d/-7d expiry), verify each lands on the correct tab via the BE filter, and check the detail-page pill matches.

## Test plan

- [x] `go build ./...` and `go test ./services/ ./apiserver/ ./registry/memory/ ./cmd/... ./models` — green
- [x] FE: `npm run typecheck && npm run lint && npm run format:check && npx vitest run` — 611 tests pass
- [x] FE: `npm run i18n:check` — clean
- [x] Migration generated via `scripts/generate-migration.sh add_warranty_tracking` against ephemeral postgres
- [x] `make lint-migrations` — verified clean against a freshly bootstrapped postgres-test (no schema drift between annotations and the generated migration)
- [x] `npx playwright test --project=chromium tests/warranties.spec.ts` — passes locally (firefox/webkit weren't run because the binaries aren't installed on this dev machine; CI runs all three)

## Decisions on the issue's "Options to decide"

1. **"Expiring" threshold:** 60 days fixed (`models.WarrantyExpiringWindowDays`). Single source of truth shared by the FE pill, the BE list filter, and the worker's first reminder threshold.
2. **Multiple warranties per item:** v1 ships as one warranty per commodity (date + notes inline on the row).
3. **Warranty document attachment:** uses the existing typed Files surface — no new file type.
4. **Reminder channels:** email only for v1.

## Out of scope (deferred follow-ups)

- Replace the Dashboard's `dashboard-warranty-placeholder` `ComingSoonBanner` with real warranty stats — the panel can light up once we expose a stats endpoint. Adjacent value-by-location/area placeholders share the same banner; replacing all three is a single dashboard polish ticket.
- Drop the legacy `warranty:YYYY-MM-DD` tag fallback in `warrantyStatus()` once a data migration has rewritten old tag-based warranties into the new field.

### CODE REVIEW

File: go/registry/postgres/commodities.go [hunk 1]
```diff
@@ -274,6 +275,46 @@ func buildCommodityWhere(opts registry.CommodityListOptions) (string, []any) {
...
+		// ISO so lexicographic comparison matches chronological order.
+		var disj []string
+		for _, s := range opts.WarrantyStatuses {
+			switch s {
+			case registry.WarrantyStatusFilterNone:
+				disj = append(disj, "(warranty_expires_at IS NULL OR warranty_expires_at = '')")
+			case registry.WarrantyStatusFilterExpired:
+				disj = append(disj, fmt.Sprintf("(warranty_expires_at IS NOT NULL AND warranty_expires_at < $%d)", idx))
+				args = append(args, today)
+				idx++
```
[the remark this answers is not in the document]
Comment (denisvmedia) [reply go/registry/postgres/commodities.go:306]: Fixed in 66209f5. All non-`none` predicates (`expired`, `expiring`, `active`) now also test `warranty_expires_at <> ''` so empty strings classify as `none` consistently with `ComputeWarrantyStatus`. Regression test: `TestCommodityRegistry_ListPaginated_WarrantyFilter_EmptyStringNotExpired`.

File: go/registry/postgres/commodities.go [hunk 2]
```diff
@@ -274,6 +275,46 @@ func buildCommodityWhere(opts registry.CommodityListOptions) (string, []any) {
...
+			}
+		}
+		if len(disj) > 0 {
+			conds = append(conds, "("+strings.Join(disj, " OR ")+")")
+		}
+	}
+	if opts.WarrantyExpiresBefore != "" {
+		conds = append(conds, fmt.Sprintf("(warranty_expires_at IS NOT NULL AND warranty_expires_at < $%d)", idx))
+		args = append(args, opts.WarrantyExpiresBefore)
```
[the remark this answers is not in the document]
Comment (denisvmedia) [reply go/registry/postgres/commodities.go:326]: Fixed in 66209f5 alongside the status predicates — `warranty_expires_before` now also guards on `warranty_expires_at <> ''`. Same regression test asserts the empty-string row doesn't leak into this filter either.

File: go/services/warranty_reminder_service.go [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+		// does not double-message after the deadline.
+		return nil
+	}
+	daysLeft := int(exp.Sub(today).Hours() / 24)
+	var out []models.WarrantyReminderThreshold
+	for _, t := range models.WarrantyReminderThresholds {
+		if daysLeft <= int(t) {
+			out = append(out, t)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (denisvmedia) [reply go/services/warranty_reminder_service.go]: Marked outdated by the next push but addressing for the record: docstring rewritten in 66209f5 to match the actual behavior (returns every threshold ≤ daysLeft so post-outage catch-up still fires the wider buckets); UTC normalisation `n := now.UTC()` added in the same commit.

File: go/services/warranty_reminder_service.go [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+	reminder := models.WarrantyReminder{
+		TenantGroupAwareEntityID: models.TenantGroupAwareEntityID{
+			TenantID:        c.TenantID,
+			GroupID:         c.GroupID,
+			CreatedByUserID: c.CreatedByUserID,
+		},
+		CommodityID:   c.ID,
+		ThresholdDays: int(threshold),
+		SentAt:        time.Now(),
+	}
```
[the remark this answers is not in the document]
Comment (denisvmedia) [reply go/services/warranty_reminder_service.go]: Fixed in f436fd0. New order in `processOne`:

1. `HasSent` fast-path — preserves the no-double-send contract.
2. Resolve recipients (fallible).
3. Try every enqueue (fallible).
4. Commit the idempotency row **only** if at least one enqueue succeeded.

If every enqueue fails the row is never written and the next sweep retries. A worker crash between (3) and (4) replays the reminder — duplicate email is the accepted trade-off vs silent loss.

`SentAt` now uses the injected sweep clock (`now`) instead of `time.Now()`. Regression test: `TestWarrantyReminderService_RemindOnce_EnqueueFailureRetries` (failing-queue → 0 rows + failed=2; rerun with working queue → both fire).

File: go/models/commodity.go [hunk 5]
```diff
@@ -146,6 +147,73 @@ type Commodity struct {
...
+	exp := expires.ToTime()
+	if exp.IsZero() {
+		return WarrantyStatusNone
+	}
+	today := time.Date(now.Year(), now.Month(), now.Day(), 0, 0, 0, 0, time.UTC)
+	if exp.Before(today) {
+		return WarrantyStatusExpired
+	}
+	cutoff := today.AddDate(0, 0, WarrantyExpiringWindowDays)
+	if !exp.After(cutoff) {
```
[the remark this answers is not in the document]
Comment (denisvmedia) [reply go/models/commodity.go:220]: Fixed in 66209f5. `ComputeWarrantyStatus` now does `n := now.UTC()` first and pulls Y/M/D from `n`. Same treatment in `matchedThresholds`. Regression test: `TestComputeWarrantyStatus_NonUTCNow` (two clocks at the same UTC instant disagreeing on local Y/M/D, asserts identical classification).

File: go/services/email_templates/warranty_reminder.txt.tmpl [hunk 6]
```diff
@@ -0,0 +1,8 @@
+Hi {{.Name}},
+
+The warranty for {{.CommodityName}} expires {{if eq .ThresholdDays 1}}tomorrow{{else}}in {{.ThresholdDays}} days{{end}} ({{.ExpiryDate}}).
```
[the remark this answers is not in the document]
Comment (denisvmedia) [reply go/services/email_templates/warranty_reminder.txt.tmpl]: Reworded in f436fd0. New copy treats `ThresholdDays` as the cadence label and surfaces the actual date separately:

> This is your {{.ThresholdDays}}-day warranty reminder for {{.CommodityName}}.
>
> The warranty expires on {{.ExpiryDate}}.

Survives a catch-up tick after worker downtime.

File: go/services/email_templates/warranty_reminder.html.tmpl [hunk 7]
```diff
@@ -0,0 +1,14 @@
+<!doctype html>
+<html lang="en">
+<body>
+<p>Hi {{.Name}},</p>
+<p>The warranty for <strong>{{.CommodityName}}</strong> expires
+{{- if eq .ThresholdDays 1}} tomorrow{{else}} in {{.ThresholdDays}} days{{end}}
+({{.ExpiryDate}}).</p>
```
[the remark this answers is not in the document]
Comment (denisvmedia) [reply go/services/email_templates/warranty_reminder.html.tmpl]: Reworded in f436fd0 in lockstep with the txt template — `<strong>{{.ThresholdDays}}-day warranty reminder</strong>` + a separate "expires on {{.ExpiryDate}}" line.

File: go/services/warranty_reminder_worker.go [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+)
+
+const defaultWarrantyReminderInterval = 1 * time.Hour
+
+// Prometheus counters for the warranty reminder worker. Registered
+// lazily at package init via promauto against the default registry.
+//
+// Labels:
+//   - threshold: "60", "30", or "7" — matches WarrantyReminderThreshold.
+//
```
[the remark this answers is not in the document]
Comment (denisvmedia) [reply go/services/warranty_reminder_worker.go:21]: Fixed in f436fd0. `RemindOnce` now returns `WarrantyReminderStats{ SentByThreshold: map[Threshold]int, Failed: int }`; the worker emits one `inventario_warranty_reminders_sent_total{threshold="60"|"30"|"7"}` series per threshold, matching the help text. The `"any"` placeholder is gone.

File: go/registry/memory/warranty_reminders.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+	r.lock.Lock()
+	for pair := r.items.Oldest(); pair != nil; pair = pair.Next() {
+		v := pair.Value
+		if v.CommodityID == reminder.CommodityID && v.ThresholdDays == reminder.ThresholdDays {
+			r.lock.Unlock()
+			return false, nil
+		}
+	}
+	r.lock.Unlock()
+	if reminder.SentAt.IsZero() {
```
[the remark this answers is not in the document]
Comment (denisvmedia) [reply go/registry/memory/warranty_reminders.go]: Marked outdated by the next push but addressing for the record: fixed in 66209f5 — `CreateOnce` now holds the lock through the entire check + insert, writing directly into `r.items` instead of delegating to `Create` (which would re-acquire the lock and force a release mid-flight). Regression test: `TestWarrantyReminderRegistry_CreateOnce_Race` runs 20 concurrent calls on the same tuple and asserts exactly one inserted=true + one row in the registry.

File: frontend/src/features/commodities/constants.ts [hunk 10]
```diff
@@ -55,30 +55,48 @@ export type CommoditySortOption = (typeof COMMODITY_SORT_OPTIONS)[number]
...
+  if (!s) return null
+  const t = Date.parse(`${s}T00:00:00Z`)
+  return Number.isNaN(t) ? null : t
+}
+
+function classifyDays(expiresAt: number): CommodityWarrantyStatus {
+  const today = Date.now()
+  const days = (expiresAt - today) / (1000 * 60 * 60 * 24)
   if (days < 0) return "expired"
   if (days <= WARRANTY_EXPIRING_DAYS) return "expiring"
```
[the remark this answers is not in the document]
Comment (denisvmedia) [reply frontend/src/features/commodities/constants.ts]: Fixed in f436fd0. `classifyDays` now anchors on UTC midnight:

```ts
const now = new Date()
const todayUTC = Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate())
const days = (expiresAt - todayUTC) / (1000 * 60 * 60 * 24)
```

Same boundary the BE's `ComputeWarrantyStatus` and the detail-page days-remaining helper use, so all four surfaces agree.

File: frontend/src/components/items/CommodityFormDialog.tsx [hunk 11]
```diff
@@ -518,21 +520,80 @@ function PurchaseStep(props: any) {
...
+// warrantyStatusFromDate is the live preview equivalent of the BE's
+// ComputeWarrantyStatus — same 60-day boundary so the form pill and
+// the list-page pill agree. Returns "none" for empty / unparseable
+// input so the preview block stays hidden.
+function warrantyStatusFromDate(d: string | undefined) {
+  if (!d) return "none" as const
+  const parsed = Date.parse(`${d}T00:00:00Z`)
+  if (Number.isNaN(parsed)) return "none" as const
+  const today = Date.now()
+  const days = (parsed - today) / (1000 * 60 * 60 * 24)
```
[the remark this answers is not in the document]
Comment (denisvmedia) [reply frontend/src/components/items/CommodityFormDialog.tsx]: Fixed in f436fd0 — `warrantyStatusFromDate` now delegates to the shared `warrantyStatus()` helper (which got the UTC-midnight fix in the same commit), so the form preview, the list pill, the detail pill and the BE filter share one boundary computation.

File: frontend/src/pages/warranties/WarrantiesListPage.tsx [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+    setSearchParams(params, { replace: true })
+  }
+
+  const rows =
+    tab === "all"
+      ? // The "all" tab includes both items with a warranty AND items
+        // without (so the user can see the universe and notice rows that
+        // need a date filling in). Filter to "has any warranty" using
+        // the same warrantyStatus() helper the row pills use, so what's
+        // counted matches what's rendered.
```
[the remark this answers is not in the document]
Comment (denisvmedia) [reply frontend/src/pages/warranties/WarrantiesListPage.tsx]: Comment updated in f436fd0 to match behavior — items without ANY warranty signal live under the dedicated "No date" tab (`status === "none"`), the "All" tab is intentionally scoped to "has any tracked warranty signal" so the user can sweep from there into per-status tabs without hopping over no-warranty rows.

File: go/services/warranty_reminder_service.go [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+		if err := s.emailSvc.SendWarrantyReminderEmail(ctx, r.email, r.name, c.Name, expiry, url, int(threshold)); err != nil {
+			// Logged but not bubbled — the email queue handles retries
+			// internally. We still count this iteration as "sent".
+			slog.Error("warranty reminder: enqueue failed",
+				"commodity_id", c.ID,
+				"to", r.email,
+				"error", err,
+			)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (denisvmedia) [reply go/services/warranty_reminder_service.go:255]: Same fix as 3199328977 (`processOne` ordering) — covered by f436fd0. If every recipient enqueue fails, the idempotency row is never written and the next sweep retries; partial success (≥1 enqueue worked) commits the row. Regression test: `TestWarrantyReminderService_RemindOnce_EnqueueFailureRetries`.

File: frontend/src/pages/commodities/CommodityDetailPage.tsx [hunk 14]
```diff
@@ -674,3 +673,85 @@ function FilesTab({
...
+          ) : null}
+        </div>
+        {daysRemaining !== null && status !== "expired" ? (
+          <p className="text-sm text-muted-foreground">
+            {daysRemaining === 0
+              ? t("commodities:detail.warranty.expiresToday")
+              : daysRemaining > 0
+                ? t("commodities:detail.warranty.daysRemaining", { count: daysRemaining })
+                : t("commodities:detail.warranty.expiredAgo", { count: -daysRemaining })}
+          </p>
```
[the remark this answers is not in the document]
Comment (denisvmedia) [reply frontend/src/pages/commodities/CommodityDetailPage.tsx]: Fixed in f436fd0 — dropped the `status !== "expired"` guard on the days-remaining paragraph. The `expiredAgo` branch is now reachable: expired warranties show "Expired N days ago" alongside the status pill.
